### DISCUSSION TITLE
Extract text

### DISCUSSION THREAD

**Discussion by lulu-ds** [jsvine/pdfplumber#773] [Ask for help with specific PDFs]
Hi there,

I used the sample codes : https://github.com/jsvine/pdfplumber/issues/242 to extract text from outside tables. But I am not able to extract text highlighted in grey (coming from the 1st grey box). 
Is there a way I can updated the table setting to extract this part of the text? Can you please help me with it? Thank you in advance!

**Comment by samkit-jain** [comment]
Hi @lulu-ds Appreciate your interest in the library. If you want to just extract the text with the grey background, you can try 
1. Filtering out the words that match a certain criteria. Like the color, size, ... 
2. Filtering out the text within the largest rectangle. My guess would be that the grey box would have been represented as a rect object and if that is the case, you can use `page.crop` to crop and extract text.

If you can share the PDF (after removing any sensitive information), I'll be able to help out better.

**Reply by samkit-jain** [reply]
Thanks for the PDF. Something like this might help
```python
import pdfplumber

pdf = pdfplumber.open("test_pdf.pdf")

page = pdf.pages[1]  # 2nd page.

# Filter out the rect objects that belong to the required area.
rects = [rect for rect in page.rects if rect["non_stroking_color"] == [0.851]]

rects = sorted(rects, key=lambda rect: (rect["x0"], rect["top"]))

final_rects = [{"x0": rects[0]["x0"], "top": rects[0]["top"], "x1": rects[0]["x1"], "bottom": rects[0]["bottom"]}]  # This will store the result.

# Combine all the rects that are closer to each other.
for rect in rects[1:]:
    distance_between_the_rects = rect["top"] - final_rects[-1]["bottom"]
    height_of_current_rect = rect["bottom"] - rect["top"]

    if distance_between_the_rects - height_of_current_rect <= 0:
        final_rects[-1]["bottom"] = rect["bottom"]
    else:
        final_rects.append({"x0": rect["x0"], "top": rect["top"], "x1": rect["x1"], "bottom": rect["bottom"]})

for rect in final_rects:
    if rect["bottom"] - rect["top"] > 10:
        cropped = page.crop((rect["x0"], rect["top"], rect["x1"], rect["bottom"]))
        print(cropped.extract_text())
```

It gives the output as 
```
1. The vendor should submit a draft of the vendor’s member incentive programs.  The vendor shall understand 
and agree that award of a contract does not constitute the state agency’s approval or acceptance of the 
vendor’s member incentive programs.  For each member incentive program proposed, the vendor should 
provide the following information: 
```
and also works on the first page.


You can take the above code as reference and optimise it to fit your needs. What I had done was amongst all the `rect` objects, find out what was different in the required area. I found that the rects in the area of interest had the `non_stroking_color` property as `[0.851]`. It would give you all the rects that have the grey background. Then, I combined the rects that are closer together to be a single rect. That is, going from
![image](https://user-images.githubusercontent.com/15127115/207398119-4d5d7469-4095-4341-8477-727edc2df0da.png)
to
![image](https://user-images.githubusercontent.com/15127115/207398000-72d3cff4-628b-4d34-945e-15103aaac1c9.png)

Then, I further narrowed it down by discarding small `rect` objects to get the areas of interest.

**Reply by lulu-ds** [reply]
Thank you [samkit-jain](https://github.com/samkit-jain)!!

And can you help me how to extract text that form the box on page 2 ? Because it is not able to extracted by using the extract_tables() or the sample codes you wrote for the text outside the table. But if I use the extract_text() which successfully extracted the text, but it ended up chopped and mixed in with the rest of the body text. Any suggestions?

**Reply by samkit-jain** [reply]
Sure, you can use the following code
```python
page = pdf.pages[1]

curves = [curve for curve in page.curves if len(curve["pts"]) == 9]

for curve in curves:
    cropped = page.crop((curve["x0"], curve["top"], curve["x1"], curve["bottom"]))
    print(cropped.extract_text())
```
It will give the output as
```
My Health Pays Top 3
Member Rewards
Top My Health Pays rewards in 2021 
(as of November): 
(cid:3511) Dental Exams: 84,060 rewards 
(over $2.1 million) distributed
(cid:3511) Annual Well Visits: 70,491 
rewards (over $1.7 million) 
distributed
(cid:3511) Flu Shots: 20,057 rewards (over 
$600,000) distributed
```
